lib/git-fastclone.rb

```diff
@@ -270,11 +270,10 @@ def update_submodules(pwd, url)
     def thread_update_submodule(submodule_url, submodule_path, threads, pwd)
       threads << Thread.new do
         with_git_mirror(submodule_url) do |mirror, _|
-          Dir.chdir(File.join(abs_clone_path, pwd).to_s) do
-            cmd = ['git', 'submodule', verbose ? nil : '--quiet', 'update', '--reference', mirror.to_s,
-                   submodule_path.to_s].compact
-            fail_on_error(*cmd, quiet: !verbose, print_on_failure: print_git_errors)
-          end
+          fail_on_error(['cd', File.join(abs_clone_path, pwd).to_s], quiet: !verbose)
+          cmd = ['cd', File.join(abs_clone_path, pwd).to_s, '&&', 'git', 'submodule',
```

Comment:
Couple things here: 
1. I don't think that running `cd` in a subshell actually changes the current working directory of the container ruby process.
```
2.7.4 :003 > BuildExecution.fail_on_error('cd', '/var/tmp')
D, [2023-08-31T17:20:32.130055 #35234] DEBUG -- build_execution: Running Shell Safe Command: "cd" "/var/tmp"

 => "" 
2.7.4 :004 > BuildExecution.fail_on_error('pwd')
D, [2023-08-31T17:20:34.590263 #35234] DEBUG -- build_execution: Running Shell Safe Command: "pwd"

/Users/jmartin/Development/ios-register-worktree
```
2. `fail_on_error` doesn't know how to process `&&` (that requires bash interpreter to interpret that boolean). I'm pretty sure this will pass the `&&` as a string literal parameter to the cd command handler. I don't know why `cd` isn't erroring out in this case, but I don't think it will do what you think it should be doing. Example:
```
2.7.4 :002 > BuildExecution.fail_on_error('echo', 'hi', '&&', 'echo', 'bye')
D, [2023-08-31T17:25:55.820719 #36253] DEBUG -- build_execution: Running Shell Safe Command: "echo" "hi" "&&" "echo" "bye"

hi && echo bye
```

I think what you want to do is pass `, chdir: <path>` in the `fail_on_error` opts arg, which will get plumbed through to `Open3.popen2e`.

Reply:
gotcha! Thx I should have tested out above myself!

 i am also thinking if doing threading here might not help improve anything at all? because git operations might still be done in serial anyway.
Testing all out now since i know the expected state of our case

Reply:
What about `git -C`? You can prepend any git command with that from what I can tell. It works like `tar -C`

<img width="597" alt="image" src="https://github.com/square/git-fastclone/assets/119358/e7f846ee-9b25-47da-ae9c-a1efd6b8a5e2">

Reply:
Good to see the “What is the way set pwd when using multithreaded ruby to write build shell scripts?” discussion is alive and well.

I recall that when fail_on_error and friends were just a prototype, the chdir option had to be plumbed in precisely because:

1) calling chdir() all over the place had dangerous threading/global state implications.
2) Not every command was nice like git and gave you an option to swap out the pwd on the command line.
3) fail_on_error was designed to avoid ever invoking the shell to cut down on unintended consequences from things like shell tokenization, and to keep the performance overhead of starting the shell to zero….

… But people mostly use the shell to change pwd…so it’s very natural to want something 

Anyway, The chdir keyword argument is there to be used, and this discussion is almost old enough to be memorizing its times tables in math class 🙂

Reply:
I think I will do both: this version (1.4.3) keeps the current chdir implementation since at high level it's the "same" as current impl just doing in the right way
And in a minor version bump (1.5.0) change it to `-C` since it now depends on git side implementation which might have other implications.